File: 0647-palindromic-substrings/0647-palindromic-substrings.java
class Solution {
    
    boolean isPalindrome(int i, int j, String S, Boolean[][] dp) {
        if (i >= j) return true;
        if (dp[i][j] == null) {
            dp[i][j] = false;
            dp[i][j] = S.charAt(i) == S.charAt(j) && isPalindrome(i + 1, j - 1, S, dp);
        }
        return dp[i][j];
    }
    
    public int countSubstrings(String s) {
        int n = s.length();
        int counter = 0;
        Boolean[][] dp = new Boolean[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (isPalindrome(i, j, s, dp)) {
                    counter++;
                }
            }
        }
        return counter;
    }
}

File: 0647-palindromic-substrings/0647-palindromic-substrings.cpp
class Solution {
public:
    
    int isPalindrome(int st, int ed, string &s) {
        if (st >= ed) return true;
        int &ret = dp[st][ed];
        if (ret != -1) return ret;
        ret = false;
        return ret = (s[st] == s[ed]) && isPalindrome(st + 1, ed - 1, s);
    }
    
    int countSubstrings(string s) {
        int n = s.size();
        int counter = 0;
        dp.resize(n, vector<int>(n, -1));
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (isPalindrome(i, j, s)) {
                    counter++;
                }
            }
        }
        return counter;
    }
private:
    vector<vector<int>> dp;
};